daemon/graphdriver/copy/copy.go

```diff
@@ -107,156 +120,393 @@ func copyXattr(srcPath, dstPath, attr string) error {
+// we have a special field for the channel that's used for this given directory
+// the reason why is to serialize writes for that dirent, and then we can simply
+// rely on delivery ordering to make sure our message to update mtimes gets there
+// last
+func (w *walkFn) walk2(ctx context.Context, path string, info os.FileInfo) error {
+	c := w.workers[w.rand.Intn(len(w.workers))]
+
+	names, err := readDirNames(path)
+	err1 := w.walkFunc(ctx, path, info, c, err)
+	// If err != nil, walk can't walk into this directory.
+	// err1 != nil means walkFn want walk to skip this directory or stop walking.
+	// Therefore, if one of err and err1 isn't nil, walk will return.
+	if err != nil || err1 != nil {
+		// The caller's behavior is controlled by the return value, which is decided
+		// by walkFn. walkFn may ignore err and return nil.
+		// If walkFn returns SkipDir, it will be handled by the caller.
+		// So walk should return whatever walkFn returns.
+		return err1
+	}
+
+	for _, name := range names {
+		filename := filepath.Join(path, name)
+		fileInfo, err := os.Lstat(filename)
+		if err != nil {
+			if err := w.walkFunc(ctx, filename, fileInfo, c, err); err != nil {
 				return err
 			}
-		}
-
-		// system.Chtimes doesn't support a NOFOLLOW flag atm
-		// nolint: unconvert
-		if f.IsDir() {
-			dirsToSetMtimes.PushFront(&dirMtimeInfo{dstPath: &dstPath, stat: stat})
-		} else if !isSymlink {
-			aTime := time.Unix(int64(stat.Atim.Sec), int64(stat.Atim.Nsec))
-			mTime := time.Unix(int64(stat.Mtim.Sec), int64(stat.Mtim.Nsec))
-			if err := system.Chtimes(dstPath, aTime, mTime); err != nil {
+		} else if fileInfo.IsDir() {
+			err = w.walk2(ctx, filename, fileInfo)
+			if err != nil {
 				return err
 			}
 		} else {
-			ts := []syscall.Timespec{stat.Atim, stat.Mtim}
-			if err := system.LUtimesNano(dstPath, ts); err != nil {
+			if err := w.walkFunc(ctx, filename, fileInfo, c, err); err != nil {
 				return err
 			}
 		}
-		return nil
-	})
+	}
+	return w.updateDirTime(ctx, path, info, c)
+}
+
+func (w *walkFn) walk(ctx context.Context, root string) error {
+	c := w.workers[w.rand.Intn(len(w.workers))]
+	info, err := os.Lstat(root)
 	if err != nil {
+		err = w.walkFunc(ctx, root, nil, c, err)
+	} else {
+		err = w.walk2(ctx, root, info)
+	}
+	return err
+}
+
+// DirCopy copies or hardlinks the contents of one directory to another,
+// properly handling xattrs, and soft links
+//
+// Copying xattrs can be opted out of by passing false for copyXattrs.
+func DirCopy(srcDir, dstDir string, copyMode Mode, copyXattrs bool) error {
+	return DirCopyWithConcurrency(srcDir, dstDir, copyMode, copyXattrs, 0)
+}
+
+// DirCopyWithConcurrency performs the same work as DirCopy, but allows you to specify the
+// concurrency level. Specifying workerCount as 0 will make it use the default worker
+// count
+func DirCopyWithConcurrency(srcDir, dstDir string, copyMode Mode, copyXattrs bool, workerCount int) error {
```

Comment:
Wondering why you're using different terminology here than in daemon/graphdriver/vfs/copy_linux.go `dirCopy()` (`workercount` vs `concurrency`). Perhaps "workercount" should be seen as an implementation detail, and this should just be named `concurrency` (after all, that's also part of the function name).

(open for input though 😅)